notary delegation integration into docker

riyazdf opened the pull request:
Adds newest notary code into docker, which includes delegations.

Docker changes:
- We prioritize a `releases` delegation role as a highest priority, but fall back to the `targets` role if the `releases` delegation role does not exist
- Update integration tests to comply with new notary error formats and server configuration format

Notary changes:
- Also introduces remote snapshot signing (the `snapshot` key will be maintained on the server) to allow for delegation functionality
- Key role information is now included in PEM headers rather than as a suffix on the key filename, but the code is backwards compatible with older keys on old repos/clients 
- Command line key rotation for a single key and/or having the server manage the snapshot key

Caveats:
- Old clients cannot push to repos created or updated by new clients
- Old clients cannot read keys on new repos created by new clients

Signed-off-by: Riyaz Faizullabhoy riyaz.faizullabhoy@docker.com

riyazdf committed:
Notary delegation integration into docker

Signed-off-by: Riyaz Faizullabhoy <riyaz.faizullabhoy@docker.com>

icecrime commented:
Moving to code review as there is no feature impact (yet) as far as I can tell.

@riyazdf Can you explain what's "WIP" about this PR?

icecrime commented on api/client/trust.go:
```diff
@@ -37,7 +37,12 @@ import (
+const releases = "releases"
+
+var (
+	releasesRole = filepath.Join(data.CanonicalTargetsRole, releases)
```
Is that actually a path, or are we using `filepath.Join` as a mean to create a `"X/Y"` string? If it's the first I'm all for it, if not then we should make it explicit, because this will produce different strings on different OS.

riyazdf replied:
Thanks for the catch, we're going for a `"X/Y"` string so I'll change that!

icecrime commented on hack/vendor.sh:
```diff
@@ -44,7 +44,9 @@ clone git github.com/boltdb/bolt v1.1.0
 
-clone git github.com/docker/notary 45de2828b5e0083bfb4e9a5a781eddb05e2ef9d0
+# get desired notary commit, might also need to be updated in Dockerfile
+clone git github.com/docker/notary ffca6fb522fc0dc9d582c848b20746e9cda9ca87
```
Just as an FYI, we're pushing more and more to get our dependencies to tag, which will make it easier for us to look into a changelog to understand changes as a we bump dependencies versions.

cyli replied:
@icecrime Do you mean tag it with the docker commit, the docker version, or both?  (asking in case we have to re-vender notary twice in the same docker version - then two commits would be tagged with the same docker version)

cyli replied:
And apologies, this was the wrong place to comment as this thread has been hidden :)

calavera commented:
it looks like @GordonTheTurtle didn't trigger anything for this PR, weird.

thaJeztah commented:
@calavera apparently that can be if there's a network hiccup between github and our CI, running `!rebuild` will start them :-)

riyazdf commented:
@icecrime this is WIP because we're aiming to add an additional cli command to notary to manage delegations, and add additional integration tests to docker/docker for these features.

cc @docker/security-team

icecrime commented:
I see. I'm going to move this back to design review as UX change are incoming.

Anything we can read to learn how this will surface in the Docker CLI?

aaronlehmann commented on vendor/src/github.com/docker/notary/tuf/data/roles.go:
```diff
@@ -97,6 +113,26 @@ func ValidRole(name string) bool {
 
+// IsDelegation checks if the role is a delegation or a root role
+func IsDelegation(role string) bool {
+	targetsBase := fmt.Sprintf("%s/", ValidRoles[CanonicalTargetsRole])
```
nit: this can be written `targetsBase = ValidRoles[CanonicalTargetsRole] + "/"`

aaronlehmann commented on vendor/src/github.com/docker/notary/tuf/data/roles.go:
```diff
@@ -97,6 +113,26 @@ func ValidRole(name string) bool {
+func IsDelegation(role string) bool {
+	targetsBase := fmt.Sprintf("%s/", ValidRoles[CanonicalTargetsRole])
+
+	whitelistedChars, err := regexp.MatchString("^[-a-z0-9_/]+$", role)
```
If `IsDelegation` is called more than infrequently, it might be a good idea to precompile the regexp into a global variable, i.e. `var delegationRegexp = regexp.MustCompile("^[-a-z0-9_/]+$")`. This also has the advantage that a syntax error in the regexp won't cause `IsDelegation` to always return false (the program would instead panic at startup).

cyli committed:
Add an integration test for docker being able to push to a repo with delegations.

Signed-off-by: cyli <cyli@twistedmatrix.com>

aaronlehmann commented on vendor/src/github.com/docker/notary/tuf/data/roles.go:
```diff
@@ -173,6 +220,76 @@ func (r Role) CheckPrefixes(hash string) bool {
+	}
+	for _, e := range new {
+		if !have[e] {
+			orig = append(orig, e)
```
I think it would be safer to return a copy of `orig` with additional items than calling `append` on `orig`. Using `append` could cause problems if the caller calls `mergeStrSlices` multiple times with the same `orig` argument, or if `orig` is a subslice of a larger slice (see http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#slice_data_corruption)

cyli commented:
@icecrime I'm not sure that any changes will surface to Docker CLI for this PR, maybe aside from a password environment variable.  The CLI changes we wanted to make were to notary client CLI so that we can use it in integration tests.

[Here is a preview of a test-in-progress we wanted to add](https://github.com/riyazdf/docker/compare/notary-delegations...cyli:notary-delegations) (this comparison view may break as this branch rebased).  It currently imports the notary library into the integration tests, in order to set up the delegation.  We'd like to add to the Notary CLI so we can just use the binary instead - the test is to make sure that in a future version (1.11?), when we add the releases delegation role and delegated keys, that this version of Docker (1.10) will be able to read from and write to it.

The test as written adds an extra environment variable for the delegation key, only because I just wasn't sure how shoehorn the password in for signing for this test.  I'm not sure we ultimately want to add that change this PR.

aaronlehmann commented on vendor/src/github.com/docker/notary/tuf/store/httpstore.go:
```diff
@@ -33,6 +45,19 @@ func (err ErrMaliciousServer) Error() string {
+
+func (err ErrInvalidOperation) Error() string {
+	if err.msg != "" {
+		return fmt.Sprintf("Trust server rejected operation: %s", err.msg)
```
Minor nit: it's customary for error messages to start with a lowercase letter, because error messages might be chained together (as is being done here).

aaronlehmann commented on vendor/src/github.com/docker/notary/tuf/tuf.go:
```diff
@@ -248,20 +349,34 @@ func (tr *Repo) InitRoot(consistent bool) error {
+// InitTargets initializes an empty targets, and returns the new empty target
+func (tr *Repo) InitTargets(role string) (*data.SignedTargets, error) {
+	r := data.Role{Name: role}
+	if !r.IsDelegation() && !(data.CanonicalRole(role) == data.CanonicalTargetsRole) {
```
`!(data.CanonicalRole(role) == data.CanonicalTargetsRole)` can be `data.CanonicalRole(role) != data.CanonicalTargetsRole`

aaronlehmann commented on vendor/src/github.com/docker/notary/tuf/validation/errors.go:
```diff
@@ -0,0 +1,126 @@
+		err = json.Unmarshal(text, &e)
+		theError = e.Error
+	default:
+		err = fmt.Errorf("do not know how to unmarshall %s", x.Name)
```
should be `unmarshal`

icecrime commented:
@cyli Thanks for clarifying, my bad for assuming that this would have more impact on Engine CLI side. Does it mean that we should proceed with code review here, or should we still wait for updates?

riyazdf commented:
To my knowledge there shouldn't be any changes to the Engine CLI side (@diogomonica and @endophage please correct me if I'm wrong), so we should be able to proceed.

Thank you @icecrime and @aaronlehmann for the helpful comments so far! :smile:

icecrime commented:
For this PR, that leaves us with the changes in `api/client/trust.go` to review on Engine side, so I'm inclined to say LGTM as this is essentially a dependency bump.

cpuguy83 commented:
LGTM.

@icecrime Your LGTM doesn't look very confident, so I'll leave to you to merge :)

cpuguy83 commented on api/client/trust.go:
```diff
@@ -37,7 +37,10 @@ import (
 
-var untrusted bool
+var (
+	releasesRole = data.CanonicalTargetsRole + "/releases"
```
minor (very minor, no need to update) nit, this should be `path.Join(data.CanonicalTargetsRole, "releases")`

endophage replied:
That's something we need to actually double check (and ensure we're consistent about throughout notary). This is a TUF role and as such will always use "/" as the path separator. I'm not sure if `path.Join` will generate the correct output for TUF on Windows (when we add support for DCT there)

thaJeztah replied:
`path.Join()` should always use a slash; https://golang.org/src/path/path.go?s=3776:3808#L140, whereas `filepath.Join()` would convert to a platform-specific separator; https://golang.org/src/path/filepath/path.go?s=5823:5855#L8

icecrime replied:
Ha! I commented just the opposite :-D https://github.com/docker/docker/pull/18887#discussion_r48383924

thaJeztah replied:
`filepath.Join()` was definitely wrong, using `path.Join()` instead of concatenation, is just a nit pick, so, all is well :+1:

icecrime commented:
Pending @aaronlehmann's comments.

thaJeztah commented:
ping @icecrime looks like the PR was updated :)

riyazdf commented:
Update on my end: we're waiting on a couple more fixes on the notary side as well as some error message improvement in docker for the notary integration.  I'll ping once these are in :)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "notary-delegations" git@github.com:riyazdf/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~5
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

riyazdf committed:
update tests and error messages, revendor notary with tag

Signed-off-by: Riyaz Faizullabhoy <riyaz.faizullabhoy@docker.com>

riyazdf commented:
removed WIP, thanks for being patient :)

The latest changes addressed the above comments, revendored notary, and updated docker with updated error messages and an integration test

icecrime commented:
LGTM

jfrazelle commented:
LGTM when green

icecrime commented:
If you don't mind squashing commits + tagging the notary release for bonus points ;-)

jfrazelle commented:
oooo ya tag, i always tag it for you all :P

cyli commented:
Apologies for repeating this, since I think I asked this in a collapsed convo, but what should we tag with? Should we tag the commit with the docker commit, the docker version, or both? (asking in case we have to re-vender notary twice in the same docker version - then two commits would be tagged with the same docker version)

jfrazelle commented:
so for example engine api tagged 0.1.3 https://github.com/docker/docker/pull/19173 and make that pr to up the version... this is a lot how it is done on runc too... so like a semver version according to what you changed

jfrazelle commented:
also checkout this https://github.com/docker/docker/pull/19053

cyli commented:
@jfrazelle Oh ok!  thanks!

riyazdf commented:
Sounds good, I can definitely squash commits (I'm thinking 3 makes most logical sense, but let me know if you prefer just 1) and handle the tag.  I'm discussing with the team for the best version tag to use and should have this wrapped up tomorrow morning :)

runcom commented:
so is this actually good to merge now?

jfrazelle commented:
Wait until the tag :)


On Jan 8, 2016, 03:44 -0800, Antonio Murdaca<notifications@github.com>, wrote:
>  
> so is this actually good to merge now?
>  
>  
> —
> Reply to this email directly orview it on GitHub(https://github.com/docker/docker/pull/18887#issuecomment-169974261).
>  
>  
>

riyazdf commented:
Added the tag `docker-v1.10-1` (we can increment the last number if we need to revendor notary for this docker 1.10 version) and squashed down into 3 commits, thanks!

jfrazelle commented:
thanks! will merge when green :)

jfrazelle commented:
may the jenkins gods be ever in your favor :smile:
